Use atomic file operations to prevent null byte injection

qiansunn opened the pull request:
There is an unexpected newline containing many null bytes (ASCII 0x00)
in the IPMB configuration section of the customer’s mlx-bf.lan.conf
file, which prevents mlx-OpenIPMI to parse the IPMB channel settings.

This unexpected newline is caused by the system power cycles during the
execution of adding ipmb configuration in file. Incomplete writes during
power loss can leave files in an inconsistent state with null bytes
padding to the expected file size. When configuration files contain
lines that begin with null bytes (ASCII 0x00) but have valid
configuration data after those null bytes, the current parser will skip
the entire line.

Replace direct file appends with atomic operations:
1. Create a temporary file
2. Copy the original configuration to the temp file
3. Append the new IPMB configuration to the temp file
4. Atomically move the temp file to replace the original

This ensures the configuration file remains in a consistent state even
if the system loses power during the update process.

RM #4608154

odrang commented on lanserv/mellanox-bf/set_emu_param.sh:
```diff
@@ -170,7 +170,13 @@ if [ "$i2cbus_slave" != "NONE" ]; then
 	fi
 
 	if ! grep -q "ipmb-$i2cbus_slave" /etc/ipmi/mlx-bf.lan.conf; then
-		echo "  ipmb 2 ipmb_dev_int /dev/ipmb-$i2cbus_slave" >> /etc/ipmi/mlx-bf.lan.conf
```
@qiansunn - do we need all the added code?

can't we just use this kind of operation:

tmpfile=$(mktemp /etc/ipmi/mlx-bf.lan.conf.XXXXXX) || exit 1
cp /etc/ipmi/mlx-bf.lan.conf "$tmpfile"
echo "  ipmb 2 ipmb_dev_int /dev/ipmb-$i2cbus_slave" >> "$tmpfile"
mv "$tmpfile" /etc/ipmi/mlx-bf.lan.conf

qiansunn replied:
Hi @odrang,
1. The cp command is not atomic on the ext4, the destination file may contain only part of the source data.
2.  ">>" operation could also cause the tmp file incomplete with null bytes.

I checked some commands like: dd, tee, cat and only the "mv" is atomic. It use rename() system call in the ext4.

odrang replied:
@qiansunn - the tmpfile is always created and the final operation is mv

qiansunn replied:
You are right, the code is upgraded

Commit:
Use atomic file operations to prevent null byte injection

There is an unexpected newline containing many null bytes (ASCII 0x00)
in the IPMB configuration section of the customer’s mlx-bf.lan.conf
file, which prevents mlx-OpenIPMI to parse the IPMB channel settings.

This unexpected newline is caused by the system power cycles during the
execution of adding ipmb configuration in file. Incomplete writes during
power loss can leave files in an inconsistent state with null bytes
padding to the expected file size. When configuration files contain
lines that begin with null bytes (ASCII 0x00) but have valid
configuration data after those null bytes, the current parser will skip
the entire line.

Replace direct file appends with atomic operations:
1. Create a temporary file
2. Copy the original configuration to the temp file
3. Append the new IPMB configuration to the temp file
4. Atomically move the temp file to replace the original

This ensures the configuration file remains in a consistent state even
if the system loses power during the update process.

RM #4608154

Signed-off-by: Qian Sun <qians@nvidia.com>

odrang commented on lanserv/mellanox-bf/set_emu_param.sh:
```diff
@@ -170,7 +170,10 @@ if [ "$i2cbus_slave" != "NONE" ]; then
+		cp /etc/ipmi/mlx-bf.lan.conf "$tmpfile" || { rm -f "$tmpfile"; exit 1; }
+		echo "  ipmb 2 ipmb_dev_int /dev/ipmb-$i2cbus_slave" >> "$tmpfile" || { rm -f "$tmpfile"; exit 1; }
+		mv "$tmpfile" /etc/ipmi/mlx-bf.lan.conf
 	fi
```
@qiansunn - I see that the code is duplicated, please consider moving to external function and add description why we are doing this in this way

qiansunn replied:
Done
